src/gui/transferlistwidget.cpp

```diff
@@ -878,17 +954,24 @@ void TransferListWidget::displayListMenu(const QPoint&)
     QList<QAction *> tagsActions;
     QMenu *tagsMenu = listMenu.addMenu(GuiIconProvider::instance()->getIcon("view-categories"), tr("Tags"));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("list-add"), tr("Add...", "Add / assign multiple tags..."));
-    tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove...", "Remove multiple tags..."));
     tagsActions << tagsMenu->addAction(GuiIconProvider::instance()->getIcon("edit-clear"), tr("Remove All", "Remove all tags"));
     tagsMenu->addSeparator();
     foreach (QString tag, tags) {
-        const bool setChecked = tagsInSelection.contains(tag);
-        tag.replace('&', "&&");  // avoid '&' becomes accelerator key
-        QAction *tagSelection = new QAction(GuiIconProvider::instance()->getIcon("inode-directory"), tag, tagsMenu);
-        tagSelection->setCheckable(true);
-        tagSelection->setChecked(setChecked);
-        tagsMenu->addAction(tagSelection);
-        tagsActions << tagSelection;
+        const Qt::CheckState initialState = tagsInAll.contains(tag) ? Qt::Checked
+                                            : tagsInAny.contains(tag) ? Qt::PartiallyChecked
+                                            : Qt::Unchecked;
+
+        const CheckBoxMenuItem::ToggleFn onToggle = [this, tag](Qt::CheckState newState)
+        {
+            if (newState == Qt::CheckState::Checked)
+                addSelectionTag(tag);
+            else if (newState == Qt::CheckState::Unchecked)
+                removeSelectionTag(tag);
+            else
+                qWarning() << "Unexpected checkstate: " << newState;
```

Comment:
You need to answer the question whether it is possible normal operation if this occurs. If not, then you need to use assertions, if Yes, then you can either ignore it completely or to output debugging message (although I still doubt its necessity, because as I understand it, the possible error here is not connected with the application logic, isn't it?). In any case, the message like "Unexpected checkstate: 2" is uninformative.